resharper/resharper-unity/src/CSharp/Feature/Services/QuickFixes/MoveQuickFixes/MonoBehaviourMoveUtil.cs

```diff
@@ -0,0 +1,263 @@
+{
+    internal static class MonoBehaviourMoveUtil
+    {
+        [CanBeNull]
+        public static IMethodDeclaration GetMonoBehaviourMethod([NotNull] IClassDeclaration classDeclaration, [NotNull] string name)
+        {
+            classDeclaration.GetPsiServices().Locks.AssertReadAccessAllowed();
+            
+            return classDeclaration.MethodDeclarations.FirstOrDefault(t => t.NameIdentifier.Name.Equals(name));
+        }
+        
+        [NotNull]
+        public static IMethodDeclaration GetOrCreateMethod([NotNull] IClassDeclaration classDeclaration, [NotNull] string methodName)
+        {
+            classDeclaration.GetPsiServices().Locks.AssertReadAccessAllowed();
+            
+            var result = GetMonoBehaviourMethod(classDeclaration, methodName);
+            if (result == null)
+            {
+                var factory = CSharpElementFactory.GetInstance(classDeclaration);
+                var declaration = (IMethodDeclaration)factory.CreateTypeMemberDeclaration("void $0(){}", methodName);
+
+                result = classDeclaration.AddClassMemberDeclarationAfter(declaration, classDeclaration.FieldDeclarations.FirstOrDefault());
+               
+            }
+            return result;
+        }
+
+        public static bool IsExpressionAccessibleInScript([NotNull]ICSharpExpression expression)
+        {
+            expression.GetPsiServices().Locks.AssertReadAccessAllowed();
+            
+            Shell.Instance.GetComponent<IShellLocks>().AssertMainThread();
+            
+            if (!expression.IsValid())
+                return false;
+            
+            var methodDeclaration = expression.GetContainingNode<IMethodDeclaration>();
+            if (methodDeclaration == null)
+                return false;
+            
+            var statement = expression.GetContainingStatementLike();
+            if (statement == null)
+                return false;
+
+            var classDeclaration = expression.GetContainingNode<IClassDeclaration>();
+            if (classDeclaration == null)
+                return false;
+
+            return IsAvailableToMoveFromMethodToMethod(expression);
+        }
+
+        /// <summary>
+        /// Is node available to move outside the loop
+        /// </summary>
+        public static bool IsAvailableToMoveFromLoop([NotNull] ITreeNode toMove, [NotNull] ILoopStatement loop)
+        {
+            toMove.GetPsiServices().Locks.AssertReadAccessAllowed();
+
+            var sourceFile = toMove.GetSourceFile();
+            if (sourceFile == null)
+                return false;
+            
+            var loopStartOffset = loop.GetTreeStartOffset().Offset;
+            return IsAvailableToMoveInner(toMove, declaredElement => 
+                    declaredElement.GetDeclarationsIn(sourceFile)
+                    .FirstOrDefault(declaration => declaration.GetSourceFile() == sourceFile)?.GetTreeStartOffset().Offset < loopStartOffset
```

Comment:
It looks like this `isElementIgnored` lambda will disallow the following transformation:
```cs
for (;;)
{
  M(Local()); // move me outside of the loop
}

void Local() { }
```
just because the local function is declared after the loop.

Wouldn't it be better to check that the declaration is not _within_ the loop instead of _before_ the loop?

JFYI: this will still break compilation in case local variables are declared but not definitely assigned before the loop. However it seems that correct check for this might be too expensive for CA/QF availability.